### ISSUE TITLE
[Troubleshoot]: FSDP Shape Error

### ISSUE THREAD

**Issue by ZZC-CN** [RLinf/RLinf#633]
### Problem description

After the SFT version was released, I tried to train with my own dataset. Initially, there were no issues. However, when the EPS announcement was published, I pulled the latest code, which caused the program to throw the error: "Expects torch.Size([1327104]) but got torch.Size([1152, 1152])" when running. Even after reverting the code, the problem persists. Could you help me identify the specific cause? The model I'm using is RLinf-Pi05-ManiSkill-25Main-SFT.

### Configuration YAML file

YAML config file:

[dataset_sft_openpi.yaml](https://github.com/user-attachments/files/24879009/dataset_sft_openpi.yaml)

[pi05.yaml](https://github.com/user-attachments/files/24879010/pi05.yaml)

[fsdp.yaml](https://github.com/user-attachments/files/24879017/fsdp.yaml)

### Log file

[run_embodiment.log](https://github.com/user-attachments/files/24879074/run_embodiment.log)

### Environment

python -V:
(openpi) root@64e20cd86179:/workspace/RLinf_backup# python -V
Python 3.11.14
uv pip list:
(openpi) root@64e20cd86179:/workspace/RLinf_backup# uv pip list
Using Python 3.11.14 environment at: /opt/venv/openpi
Package                            Version     Editable project location
---------------------------------- ----------- -------------------------
absl-py                            2.3.1
accelerate                         1.12.0
accessible-pygments                0.0.5
aiohappyeyeballs                   2.6.1
aiohttp                            3.13.2
aiohttp-cors                       0.8.1
aiosignal                          1.4.0
alabaster                          1.0.0
annotated-types                    0.7.0
antlr4-python3-runtime             4.11.1
anyio                              4.9.0
argcomplete                        3.6.3
arm-pytorch-utilities              0.4.3
array-record                       0.8.3
asttokens                          3.0.1
astunparse                         1.6.3
attrs                              25.4.0
augmax                             0.4.1
av                                 16.0.1
babel                              2.17.0
backrefs                           6.1
bddl                               3.6.0
beartype                           0.19.0
beautifulsoup4                     4.13.4
blinker                            1.9.0
boto                               2.49.0
cachetools                         5.5.2
certifi                            2025.6.15
cffi                               2.0.0
cfgv                               3.5.0
charset-normalizer                 3.4.2
chex                               0.1.90
click                              8.2.1
cloudpickle                        3.1.2
cmake                              4.2.0
colorama                           0.4.6
colorful                           0.6.0a1
commonmark                         0.9.1
contourpy                          1.3.3
crcmod                             1.7
cryptography                       43.0.3
cycler                             0.12.1
dacite                             1.9.2
datasets                           3.6.0
debugpy                            1.8.19
decorator                          5.2.1
deepdiff                           8.6.1
deprecated                         1.3.1
diffusers                          0.36.0
dill                               0.3.8
distlib                            0.4.0
dm-control                         1.0.34
dm-env                             1.6
dm-tree                            0.1.9
docstring-parser                   0.17.0
docutils                           0.21.2
donfig                             0.8.1.post1
draccus                            0.10.0
easydict                           1.13
einops                             0.8.1
equinox                            0.13.2
etils                              1.13.0
evdev                              1.9.2
exceptiongroup                     1.3.0
executing                          2.2.1
farama-notifications               0.0.4
fast-kinematics                    0.2.2
fasteners                          0.20
fastjsonschema                     2.21.2
filelock                           3.20.1
flash-attn                         2.7.4.post1
flask                              3.1.2
flatbuffers                        25.9.23
flax                               0.10.2
fonttools                          4.61.1
frozenlist                         1.8.0
fsspec                             2025.3.0
future                             1.0.0
gast                               0.7.0
gcs-oauth2-boto-plugin             3.3
gcsfs                              2025.3.0
gdown                              5.2.0
ghp-import                         2.1.0
gitdb                              4.0.12
gitpython                          3.1.45
glfw                               2.10.0
google-api-core                    2.28.1
google-apitools                    0.5.35
google-auth                        2.39.0
google-auth-httplib2               0.3.0
google-auth-oauthlib               1.2.3
google-cloud-core                  2.5.0
google-cloud-storage               3.7.0
google-crc32c                      1.8.0
google-pasta                       0.2.0
google-reauth                      0.1.1
google-resumable-media             2.8.0
googleapis-common-protos           1.72.0
grpcio                             1.67.1
gsutil                             5.35
gym                                0.26.2
gym-aloha                          0.1.3
gym-notices                        0.1.0
gymnasium                          0.29.1
h11                                0.16.0
h5py                               3.15.1
hf-transfer                        0.1.9
hf-xet                             1.2.1rc0
httpcore                           1.0.9
httplib2                           0.20.4
httpx                              0.28.1
huggingface-hub                    0.36.0
humanize                           4.14.0
hydra-core                         1.4.0.dev1
icmplib                            3.0.4
identify                           2.6.15
idna                               3.10
imageio                            2.37.2
imageio-ffmpeg                     0.6.0
imagesize                          1.4.1
immutabledict                      4.2.2
importlib-metadata                 8.0.0
importlib-resources                6.5.2
iniconfig                          2.3.0
inquirerpy                         0.3.4
ipython                            9.8.0
ipython-pygments-lexers            1.1.1
itsdangerous                       2.2.0
jax                                0.5.3
jax-cuda12-pjrt                    0.5.3
jax-cuda12-plugin                  0.5.3
jaxlib                             0.5.3
jaxtyping                          0.2.36
jedi                               0.19.2
jinja2                             3.1.6
joblib                             1.5.3
jsonlines                          4.0.0
jsonschema                         4.25.1
jsonschema-specifications          2025.9.1
jupyter-core                       5.9.1
jupytext                           1.19.0.dev0
keras                              3.13.0
kiwisolver                         1.4.10rc0
labmaze                            1.0.6
latex2sympy2                       1.9.0
lerobot                            0.1.0
libclang                           18.1.1
libero                             0.1.0       /opt/venv/openpi/libero
liger-kernel                       0.6.4
llvmlite                           0.44.0
lxml                               6.0.2
mani-skill                         3.0.0b22
markdown                           3.8.2
markdown-it-py                     3.0.0
markupsafe                         3.0.2
matplotlib                         3.10.8
matplotlib-inline                  0.2.1
mdit-py-plugins                    0.4.2
mdurl                              0.1.2
mergedeep                          1.3.4
mkdocs                             1.6.1
mkdocs-get-deps                    0.2.0
mkdocs-material                    9.7.1
mkdocs-material-extensions         1.3.1
ml-collections                     1.0.0
ml-dtypes                          0.5.4
monotonic                          1.6
mplib                              0.1.1
mpmath                             1.3.0
msgpack                            1.1.2
mujoco                             3.4.0
multidict                          6.7.0
multiprocess                       0.70.16
mypy-extensions                    1.1.0
myst-parser                        4.0.1
namex                              0.1.0
nbformat                           5.10.4
nest-asyncio                       1.6.0
networkx                           3.6.1
ninja                              1.13.0
nltk                               3.9.2
nodeenv                            1.9.1
numba                              0.61.2
numcodecs                          0.16.5
numpy                              1.26.4
numpydantic                        1.7.0
nvidia-cublas-cu12                 12.4.5.8
nvidia-cuda-cupti-cu12             12.4.127
nvidia-cuda-nvcc-cu12              12.9.86
nvidia-cuda-nvrtc-cu12             12.4.127
nvidia-cuda-runtime-cu12           12.4.127
nvidia-cudnn-cu12                  9.1.0.70
nvidia-cufft-cu12                  11.2.1.3
nvidia-curand-cu12                 10.3.5.147
nvidia-cusolver-cu12               11.6.1.9
nvidia-cusparse-cu12               12.3.1.170
nvidia-cusparselt-cu12             0.6.2
nvidia-ml-py                       13.580.82
nvidia-nccl-cu12                   2.21.5
nvidia-nvjitlink-cu12              12.4.127
nvidia-nvtx-cu12                   12.4.127
nvitop                             1.6.1
oauth2client                       4.1.3
oauthlib                           3.3.1
omegaconf                          2.4.0.dev4
opencensus                         0.11.4
opencensus-context                 0.2.dev0
opencv-python                      4.11.0.86
opencv-python-headless             4.11.0.86
openexr                            3.4.4
openpi                             0.1.0
openpi-client                      0.1.0
opentelemetry-api                  1.26.0
opentelemetry-exporter-prometheus  0.47b0
opentelemetry-proto                1.39.1
opentelemetry-sdk                  1.26.0
opentelemetry-semantic-conventions 0.47b0
opt-einsum                         3.4.0
optax                              0.2.6
optree                             0.18.0
orbax-checkpoint                   0.11.13
orderly-set                        5.5.0
orjson                             3.11.5
packaging                          25.0
paginate                           0.5.7
pandas                             3.0.0rc0
parso                              0.8.5
pathspec                           1.0.3
peft                               0.11.1
pexpect                            4.9.0
pfzy                               0.3.4
pillow                             12.0.0
pip                                25.3
platformdirs                       4.5.1
pluggy                             1.6.0
polars                             1.36.1
polars-runtime-32                  1.36.1
pre-commit                         4.5.1
prettytable                        3.17.0
prometheus-client                  0.23.1
promise                            2.3
prompt-toolkit                     3.0.52
propcache                          0.4.1
proto-plus                         1.27.0
protobuf                           6.33.2
psutil                             7.1.3
ptyprocess                         0.7.0
pure-eval                          0.2.3
py-spy                             0.4.1
pyarrow                            22.0.0
pyasn1                             0.6.1
pyasn1-modules                     0.4.2
pybind11                           3.0.1
pycparser                          2.23
pydantic                           2.12.5
pydantic-core                      2.41.5
pydata-sphinx-theme                0.16.1
pyecharts                          2.0.9
pygments                           2.19.2
pylatexenc                         3.0a33
pymdown-extensions                 10.20
pymunk                             7.2.0
pynput                             1.8.1
pyopengl                           3.1.10
pyopenssl                          24.2.1
pyparsing                          3.3.0b1
pyperclip                          1.11.0
pyrealsense2                       2.56.5.9235
pysocks                            1.7.1
pytest                             9.0.2
python-dateutil                    2.9.0.post0
python-xlib                        0.33
pytorch-kinematics                 0.7.6
pytorch-seed                       0.2.0
pyu2f                              0.1.5
pyvers                             0.1.0
pyyaml                             6.0.2
pyyaml-env-tag                     1.1
pyyaml-include                     1.4.1
pyzmq                              27.1.0
ray                                2.47.1
recommonmark                       0.7.1
referencing                        0.37.0
regex                              2025.11.3
requests                           2.32.4
requests-oauthlib                  2.0.0
rerun-sdk                          0.23.1
retry-decorator                    2.0a1
rich                               14.2.0
robosuite                          1.4.1
roman-numerals                     4.1.0
rpds-py                            0.30.0
rsa                                4.7.2
ruff                               0.14.9
safetensors                        0.7.0
sapien                             3.0.1
scipy                              1.17.0rc1
sentencepiece                      0.2.1
sentry-sdk                         3.0.0a7
setuptools                         75.8.2
simple-parsing                     0.1.7
simplejson                         3.20.2
six                                1.17.0
smart-open                         7.5.0
smmap                              5.0.2
sniffio                            1.3.1
snowballstemmer                    3.0.1
soupsieve                          2.7
sphinx                             8.1.3
sphinx-autobuild                   2024.10.3
sphinx-copybutton                  0.5.2
sphinx-last-updated-by-git         0.3.8
sphinx-markdown-tables             0.0.17
sphinx-rtd-theme                   3.0.2
sphinx-sitemap                     2.7.2
sphinxcontrib-applehelp            2.0.0
sphinxcontrib-devhelp              2.0.0
sphinxcontrib-htmlhelp             2.1.0
sphinxcontrib-jquery               4.1
sphinxcontrib-jsmath               1.0.1
sphinxcontrib-qthelp               2.0.0
sphinxcontrib-serializinghtml      2.0.0
sphinxcontrib-video                0.4.1
stack-data                         0.6.3
starlette                          0.46.2
svgwrite                           1.4.3
swanlab                            0.7.4
sympy                              1.13.1
tabulate                           0.9.0
tensorboard                        2.20.0
tensorboard-data-server            0.7.2
tensordict                         0.10.0
tensorflow                         2.20.0
tensorflow-addons                  0.23.0
tensorflow-datasets                4.9.9
tensorflow-graphics                2021.12.3
tensorflow-metadata                1.17.2
tensorstore                        0.1.80
termcolor                          3.2.0
timm                               0.9.10
tokenizers                         0.21.4
toml                               0.10.2
tomli                              2.2.1
toolz                              1.1.0
toppra                             0.6.3
torch                              2.6.0
torch-memory-saver                 0.0.9
torchcodec                         0.2.0
torchdata                          0.11.0
torchvision                        0.21.0
tqdm                               4.67.1
tqdm-loggable                      0.2
traitlets                          5.14.3
transformers                       4.53.2
transforms3d                       0.4.2
tree                               0.2.4
treescope                          0.1.10
trimesh                            4.10.1
triton                             3.2.0
typeguard                          4.4.4
typing-extensions                  4.15.0
typing-inspect                     0.9.0
typing-inspection                  0.4.2
tyro                               1.0.2
tzdata                             2025.3
urllib3                            2.5.0
uvicorn                            0.35.0
virtualenv                         20.35.4
wadler-lindig                      0.1.7
wandb                              0.23.1
watchdog                           6.0.0
watchfiles                         1.1.0
wcwidth                            0.2.14
websockets                         15.0.1
werkzeug                           3.1.4
wheel                              0.45.1
word2number                        1.1
wrapt                              2.0.1
xxhash                             3.6.0
yarl                               1.22.0
zarr                               3.1.5
zipp                               3.23.0
nvidia-smi:
(openpi) root@64e20cd86179:/workspace/RLinf_backup# nvidia-smi
Tue Jan 27 09:18:36 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 570.133.20             Driver Version: 570.133.20     CUDA Version: 12.8     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A100 80GB PCIe          On  |   00000000:16:00.0 Off |                    0 |
| N/A   35C    P0             98W /  300W |   25289MiB /  81920MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA A100 80GB PCIe          On  |   00000000:27:00.0 Off |                    0 |
| N/A   26C    P0             47W /  300W |       4MiB /  81920MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   2  NVIDIA A100 80GB PCIe          On  |   00000000:38:00.0 Off |                    0 |
| N/A   39C    P0            101W /  300W |    3727MiB /  81920MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   3  NVIDIA A100 80GB PCIe          On  |   00000000:5A:00.0 Off |                    0 |
| N/A   54C    P0            120W /  300W |    7500MiB /  81920MiB |     63%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   4  NVIDIA A100 80GB PCIe          On  |   00000000:98:00.0 Off |                    0 |
| N/A   27C    P0             46W /  300W |       4MiB /  81920MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   5  NVIDIA A100 80GB PCIe          On  |   00000000:A8:00.0 Off |                    0 |
| N/A   28C    P0             47W /  300W |       4MiB /  81920MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   6  NVIDIA A100 80GB PCIe          On  |   00000000:B8:00.0 Off |                    0 |
| N/A   38C    P0             97W /  300W |   25597MiB /  81920MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   7  NVIDIA A100 80GB PCIe          On  |   00000000:D8:00.0 Off |                    0 |
| N/A   26C    P0             46W /  300W |       4MiB /  81920MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
+-----------------------------------------------------------------------------------------+
RLinf version:
RLinf version: 0.2.0
Docker image tag: 
rlinf/rlinf:agentic-rlinf0.1-torch2.6.0-openvla-openvlaoft-pi0

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by ZZC-CN** [comment]
I have identified the cause of the error.

**Root Cause**

When `use_orig_params: True`, FSDP requires all parameters to be correctly wrapped. The issue lies in:

**Incomplete FSDP wrap policy**

`_no_split_modules` and `_no_split_names` do not include all modules that need to be wrapped.

The following modules are not wrapped:

- `SiglipEncoderLayer` (vision encoder layers, parameter shape `[1152, 1152]`)
- `PaliGemmaMultiModalProjector` (multi-modal projector, parameter shape `[2048, 1152]`)
- `embed_tokens` (embedding layer, parameter shape `[257152, 2048]`)

**FSDP writeback mechanism**

- Wrapped parameters are flattened (e.g., `[1152, 1152]` → `[1327104]`).
- During forward pass, FSDP writes the flattened parameters back to their original shape.
- Unwrapped parameters remain in their original shape, leading to shape mismatches.

**Parameter view issue (secondary)**

- `to_bfloat16_for_selected_params` may create parameter views.
- View parameters can cause issues in FSDP.

**Why this problem occurs:**

The wrap policy in `OpenPi0ForRLActionPrediction._no_split_modules` only includes transformer decoder layers (`GemmaMLP` or `GemmaDecoderLayer`) but not the vision encoder layers (`SiglipEncoderLayer`). Similarly, the multi-modal projector and embedding layer are not explicitly included in the wrap policy. When FSDP wraps the model, it only wraps modules matching the policy, leaving these critical modules unwrapped. During forward pass, FSDP attempts to writeback flattened parameters to their original shape, but encounters unwrapped parameters still in their original shape, causing the shape mismatch error.

**Solution:**

Add the missing modules to the wrap policy

**I don't know why this problem occurs.**

<img width="1283" height="849" alt="Image" src="https://github.com/user-attachments/assets/df03fef2-4e95-42db-bfab-66c7da180fcf" />
